org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/UiPlugin.java

```diff
@@ -127,16 +129,35 @@ private void registerListeners() {
         CorePlugin.listenerRegistry().addEventListener(this.workingSetsAddingProjectCreatedListener);
 
         this.contextActivatingSelectionListener = new ContextActivatingSelectionListener(UiPluginConstants.GRADLE_NATURE_CONTEXT_ID, Predicates.hasGradleNature(), getWorkbench());
-        ((ISelectionService) getWorkbench().getActiveWorkbenchWindow().getService(ISelectionService.class)).addSelectionListener(this.contextActivatingSelectionListener);
+        Display display = PlatformUI.getWorkbench().getDisplay();
+        if (display != null && !display.isDisposed()) {
+            display.asyncExec(new Runnable() {
+
+                @Override
+                public void run() {
+                    ((ISelectionService) getWorkbench().getActiveWorkbenchWindow().getService(ISelectionService.class))
+                    .addSelectionListener(UiPlugin.this.contextActivatingSelectionListener);
+                }
+            });
+        }
     }
 
     @SuppressWarnings({"cast", "RedundantCast"})
     private void unregisterListeners() {
-        // if the selection service is disposed, then the listeners are already removed
-        // (see the javadoc on ISelectionService.addSelectionListener(ISelectionListener listener))
-        ISelectionService selectionService = ((ISelectionService) getWorkbench().getService(ISelectionService.class));
-        if (selectionService != null) {
-            selectionService.removeSelectionListener(this.contextActivatingSelectionListener);
+        Display display = PlatformUI.getWorkbench().getDisplay();
+        if (display != null && !display.isDisposed()) {
+            display.asyncExec(new Runnable() {
+
+                @Override
+                public void run() {
+                     // if the selection service is disposed, then the listeners are already removed
+                    // (see the javadoc on ISelectionService.addSelectionListener(ISelectionListener listener))
+                    ISelectionService selectionService = ((ISelectionService) getWorkbench().getService(ISelectionService.class));
```

Comment:
Thanks for pointing at this. Actually every WorkbenchWindow or MWindow(for e4 Apps) has its own ISelectionService. This means that the ISelectionService should be obtained from the workbenchwindow's getService(...) method rather than from the workbench itself, like I've done it in the registerListeners() method of the UIPlugin class.